ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/Witnesses.scala

```diff
@@ -10,86 +10,94 @@ import scalaz.Tag
 
 final class Witnesses(session: LedgerSession) extends LedgerTestSuite(session) {
 
-  val respectDisclosureRules =
+  private[this] val respectDisclosureRules =
     LedgerTest("RespectDisclosureRules", "The ledger should respect disclosure rules") {
       implicit context =>
         for {
           Vector(alice, bob, charlie) <- allocateParties(3)
 
-          // Create Witnesses contract
-          (witnessTxId, witnesses) <- create(WitnessesTemplate(alice, bob, charlie))(alice)
-          witnessTx <- transactionTreeById(witnessTxId, alice, bob, charlie)
+          // Create the Witnesses contract as Alice and get the resulting transaction as seen by all parties
+          (witnessesTransactionId, witnesses) <- createAndGetTransactionId(
+            WitnessesTemplate(alice, bob, charlie))(alice)
+          witnessesTransaction <- transactionTreeById(witnessesTransactionId, alice, bob, charlie)
 
-          // Divulge Witnesses contract to charlie, who's just an actor and thus cannot
-          // see it by default.
-          (_, divulgeWitness) <- create(DivulgeWitnesses(alice, charlie))(charlie)
-          _ <- exercise(divulgeWitness.contractId.exerciseDivulge(alice, witnesses.contractId))(
-            alice)
+          // Charlie is not a stakeholder of Witnesses and thus cannot see any such contract unless divulged.
+          // Such contract is divulged by creating a DivulgeWitness with Charlie as a stakeholder and exercising
+          // a choice as Alice that causes divulgence (in this case, the Witnesses instance previously
+          // created is fetched as part of the transaction).
+          divulgeWitness <- create(DivulgeWitnesses(alice, charlie))(charlie)
+          _ <- exercise(divulgeWitness.contractId.exerciseDivulge(_, witnesses.contractId))(alice)
 
-          // Now, first try the non-consuming choice
-          nonConsumingTxId <- exercise(
-            witnesses.contractId.exerciseWitnessesNonConsumingChoice(charlie))(charlie)
-          nonConsumingTx <- transactionTreeById(nonConsumingTxId, alice, bob, charlie)
+          // A non-consuming choice is exercised with the expectation
+          // that this will not cause the contract to be divulged
```

Comment:
At this point the `Witnesses` contract already has been divulged. This tests that `Charlie` can exercise a choice on the divulged contract.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2490/commits/8d4c15b70a1fd06cdd0549950ae0f0005b80f823